erts/emulator/beam/erl_bif_coverage.c

```diff
@@ -0,0 +1,315 @@
+    modp = erts_get_module(BIF_ARG_1, code_ix);
+    if (modp == NULL) {
+        BIF_ERROR(BIF_P, BADARG);
+    }
+    hdr = modp->curr.code_hdr;
+    if (hdr == NULL) {
+        BIF_ERROR(BIF_P, BADARG);
+    }
+
+    switch (hdr->coverage_mode) {
+    case ERTS_COV_FUNCTION_COVERAGE: {
+        byte* p = hdr->coverage;
+
+        res = NIL;
+        alloc_size = (3 + 3 + 2) * hdr->num_functions;
+        hp = HAlloc(BIF_P, alloc_size);
+        hend = hp + alloc_size;
+        for (i = (ssize_t)(hdr->num_functions - 1); i >= 0; i--) {
+            if (hdr->functions[i]->mfa.function == am_module_info &&
+                (hdr->functions[i]->mfa.arity == 0 ||
+                 hdr->functions[i]->mfa.arity == 1)) {
+                /* For consistency with function coverage in line
+                 * counters mode, don't include module_info/0 and
+                 * module_info/1 in the list. */
+                continue;
+            }
+
+            tmp = TUPLE2(hp, hdr->functions[i]->mfa.function,
+                         make_small(hdr->functions[i]->mfa.arity));
+            hp += 3;
+            tmp = TUPLE2(hp, tmp, p[i] ? am_true : am_false);
+            hp += 3;
+            res = CONS(hp, tmp, res);
+            hp += 2;
+        }
+        HRelease(BIF_P, hend, hp);
+        BIF_RET(res);
+    }
+    case ERTS_COV_LINE_COVERAGE:
+    case ERTS_COV_LINE_COUNTERS: {
+        const BeamCodeLineTab *lt = hdr->line_table;
+        Eterm covered;
+
+        res = NIL;
+        alloc_size = (3 + 3 + 2) * hdr->num_functions;
+        hp = HAlloc(BIF_P, alloc_size);
+        hend = hp + alloc_size;
+        for (i = (ssize_t)(hdr->num_functions - 1); i >= 0; i--) {
+            int index = lt->func_tab[i] - lt->func_tab[0];
+            int high = lt->func_tab[i+1] - lt->func_tab[0];
+
+            if (hdr->functions[i]->mfa.function == am_module_info &&
+                (hdr->functions[i]->mfa.arity == 0 ||
+                 hdr->functions[i]->mfa.arity == 1)) {
+                /* There are no executeable_line instruction in the
+                 * module_info/0 and module_info/1 functions, and
+                 * therefore they would never be shown to be covered.
+                 * Better not include in them in the list. */
+                continue;
+            }
+
+            if (hdr->coverage_mode == ERTS_COV_LINE_COVERAGE) {
+                byte* p = hdr->coverage;
+                while (index < high) {
+                    if (p[index]) {
+                        break;
+                    }
+                    index++;
+                }
+            } else if (hdr->coverage_mode == ERTS_COV_LINE_COUNTERS) {
+                Uint* p = hdr->coverage;
+                while (index < high) {
+                    if (p[index]) {
+                        break;
+                    }
+                    index++;
+                }
+            } else {
+                ASSERT(0);
+            }
```

Comment:
This looks pretty verbose, how about?
```suggestion
            if (hdr->coverage_mode == ERTS_COV_LINE_COVERAGE) {
                for (const byte *p = hdr->coverage; index < high && !p[index];) {
                    index++
                }
            } else {
                ASSERT(hdr->coverage_mode == ERTS_COV_LINE_COUNTERS);
                for (const Uint *p = hdr->coverage; index < high && !p[index];) {
                    index++
                }
            }
```